File: src/intviewer/datapresentation.cc
/*
    LibGViewer - GTK+ File Viewer library
    Copyright (C) 2006 Assaf Gordon

    Part of
        GNOME Commander - A GNOME based file manager
        Copyright (C) 2001-2006 Marcus Bjurman
        Copyright (C) 2007-2009 Piotr Eljasiak

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation; either version 2 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
*/

/*
 TODO:    change "eol" to "next_line_offset"
*/

#include <config.h>
#include <glib.h>
#include <string.h>
#include <stdlib.h>
#include <stdio.h>
#include "gvtypes.h"

#include "inputmodes.h"
#include "datapresentation.h"

using namespace std;


typedef offset_type (*align_offset_to_line_start_proc)(GVDataPresentation *dp, offset_type offset);
typedef offset_type (*scroll_lines_proc)(GVDataPresentation *dp, offset_type current_offset, int delta);
typedef offset_type (*get_end_of_line_offset_proc)(GVDataPresentation *dp, offset_type start_of_line);

struct GVDataPresentation
{
    GVInputModesData *imd;
    guint wrap_limit;
    guint fixed_count;
    offset_type max_offset;
    guint tab_size;

    PRESENTATION presentation_mode;

    align_offset_to_line_start_proc align_offset_to_line_start;
    scroll_lines_proc               scroll_lines;
    get_end_of_line_offset_proc     get_end_of_line_offset;
};

static offset_type nowrap_align_offset(GVDataPresentation *dp, offset_type offset);
static offset_type nowrap_scroll_lines(GVDataPresentation *dp, offset_type current_offset, int delta);
static offset_type nowrap_get_eol(GVDataPresentation *dp, offset_type start_of_line);

static offset_type wrap_align_offset(GVDataPresentation *dp, offset_type offset);
static offset_type wrap_scroll_lines(GVDataPresentation *dp, offset_type current_offset, int delta);
static offset_type wrap_get_eol(GVDataPresentation *dp, offset_type start_of_line);

static offset_type binfixed_align_offset(GVDataPresentation *dp, offset_type offset);
static offset_type binfixed_scroll_lines(GVDataPresentation *dp, offset_type current_offset, int delta);
static offset_type binfixed_get_eol(GVDataPresentation *dp, offset_type start_of_line);


/*********************************************************
   Data presentation public functions
*********************************************************/
GVDataPresentation *gv_data_presentation_new()
{
    return g_new0(GVDataPresentation, 1);
}


void gv_init_data_presentation(GVDataPresentation *dp, GVInputModesData *imd, offset_type max_offset)
{
    g_return_if_fail (dp!=NULL);
    g_return_if_fail (imd!=NULL);

    memset(dp, 0, sizeof(GVDataPresentation));
    dp->imd = imd;
    dp->max_offset = max_offset;
    dp->tab_size = 8;

    gv_set_data_presentation_mode(dp, PRSNT_NO_WRAP);
}


void gv_free_data_presentation(GVDataPresentation *dp)
{
}


void gv_set_data_presentation_mode(GVDataPresentation *dp, PRESENTATION present)
{
    g_return_if_fail (dp!=NULL);
    dp->presentation_mode = present;

    switch (present)
    {
        case PRSNT_NO_WRAP:
            dp->align_offset_to_line_start = nowrap_align_offset;
            dp->scroll_lines = nowrap_scroll_lines;
            dp->get_end_of_line_offset = nowrap_get_eol;
            break;

        case PRSNT_WRAP:
            dp->align_offset_to_line_start = wrap_align_offset;
            dp->scroll_lines = wrap_scroll_lines;
            dp->get_end_of_line_offset = wrap_get_eol;
            break;

        case PRSNT_BIN_FIXED:
            dp->align_offset_to_line_start = binfixed_align_offset;
            dp->scroll_lines = binfixed_scroll_lines;
            dp->get_end_of_line_offset = binfixed_get_eol;
            break;
    }
}


PRESENTATION gv_get_data_presentation_mode(GVDataPresentation *dp)
{
    g_return_val_if_fail (dp!=NULL, PRSNT_NO_WRAP);
    return dp->presentation_mode;
}


void gv_set_tab_size(GVDataPresentation *dp, guint tab_size)
{
    g_return_if_fail (dp!=NULL);
    dp->tab_size = tab_size;
}


void gv_set_fixed_count(GVDataPresentation *dp, guint chars_per_line)
{
    g_return_if_fail (dp!=NULL);
    dp->fixed_count = chars_per_line;
}


void gv_set_wrap_limit(GVDataPresentation *dp, guint chars_per_line)
{
    g_return_if_fail (dp!=NULL);
    dp->wrap_limit= chars_per_line;
}


offset_type gv_align_offset_to_line_start(GVDataPresentation *dp, offset_type offset)
{
    g_return_val_if_fail (dp!=NULL, 0);
    g_return_val_if_fail (dp->align_offset_to_line_start!=NULL, 0);
    return dp->align_offset_to_line_start(dp, offset);
}


offset_type gv_scroll_lines(GVDataPresentation *dp, offset_type current_offset, int delta)
{
    g_return_val_if_fail (dp!=NULL, 0);
    g_return_val_if_fail (dp->scroll_lines!=NULL, 0);
    return dp->scroll_lines(dp, current_offset, delta);
}


offset_type gv_get_end_of_line_offset(GVDataPresentation *dp, offset_type start_of_line)
{
    g_return_val_if_fail (dp!=NULL, 0);
    g_return_val_if_fail (dp->get_end_of_line_offset!=NULL, 0);
    return dp->get_end_of_line_offset(dp, start_of_line);
}


/***********************************************************************
  Data presentation specific implementations
***********************************************************************/
/*
 scans the file from offset "start" backwards, until a CR/LF is found.
 returns the offset of the previous CR/LF, or 0 (if we've reached the start of the file)
*/
static offset_type find_previous_crlf(GVDataPresentation *dp, offset_type start)
{
    offset_type offset = start;

    while (TRUE)
    {
        if (offset<=0)
            return 0;

        offset = gv_input_get_previous_char_offset(dp->imd, offset);
        char_type value = gv_input_mode_get_utf8_char(dp->imd, offset);

        if (value==INVALID_CHAR)
            break;

        // break upon end of line
        if (value=='\n' || value=='\r')
            break;
    }

    return offset;
}


static offset_type nowrap_align_offset(GVDataPresentation *dp, offset_type offset)
{
    while (offset>0)
    {
        char_type value = gv_input_mode_get_utf8_char(dp->imd, offset);
        if (value==INVALID_CHAR)
            return 0;
        if (value=='\r' || value=='\n')
            break;
        offset = gv_input_get_previous_char_offset(dp->imd, offset);
    }
    if (offset>0)
        return gv_input_get_next_char_offset(dp->imd, offset);
    return 0;
}


static offset_type nowrap_scroll_lines(GVDataPresentation *dp, offset_type current_offset, int delta)
{
    gboolean forward = TRUE;

    if (delta==0)
        return current_offset;

    if (delta<0)
    {
        delta = abs(delta);
        forward = FALSE;
    }
    while (delta--)
    {
        offset_type temp;

        if (forward)
            temp = nowrap_get_eol(dp, current_offset);
        else
        {
            // We need TWO CR/LF. this is not an error
            temp = find_previous_crlf(dp, current_offset);
            temp = find_previous_crlf(dp, temp);
            if (temp>0)
                temp = gv_input_get_next_char_offset(dp->imd, temp);
        }

        // Offset didn't changed ? we've reached eof
        if (temp==current_offset)
            break;

        current_offset = temp;
    }

    return current_offset;
}


static offset_type nowrap_get_eol(GVDataPresentation *dp, offset_type start_of_line)
{
    offset_type offset = start_of_line;

    while (TRUE)
    {
        char_type value = gv_input_mode_get_utf8_char(dp->imd, offset);

        if (value==INVALID_CHAR)
            break;

        offset = gv_input_get_next_char_offset(dp->imd, offset);

        // break upon end of line
        if (value=='\n' || value=='\r')
            break;
    }

    return offset;
}


/*
    returns the start offset of the previous line,
    with special handling for wrap mode.
*/
static offset_type find_previous_wrapped_text_line(GVDataPresentation *dp, offset_type start)
{
    offset_type offset = start;

    /* step 1:
        find TWO previous CR/LF = start offset of previous text line
    */
    offset = find_previous_crlf(dp, offset);
    offset = find_previous_crlf(dp, offset);
    if (offset>0)
        offset = gv_input_get_next_char_offset(dp->imd, offset);

    /* Step 2
    */

    while (TRUE)
    {
        offset_type next_line_offset = wrap_get_eol (dp, offset);

        // this is the line we want: When the next line's offset is the current
        // offset ('start' parameter), 'offset' will point to the previous
        // displayable line

        if (next_line_offset>=start)
            return offset;

        offset = next_line_offset;
    }

    return 0;  // should never get here
}


static offset_type wrap_align_offset(GVDataPresentation *dp, offset_type offset)
{
    offset_type line_start = nowrap_align_offset(dp, offset);

    for (offset_type temp=line_start; temp<=offset; temp=wrap_scroll_lines(dp, temp, 1))
        line_start = temp;

    return line_start;
}


static offset_type wrap_scroll_lines(GVDataPresentation *dp, offset_type current_offset, int delta)
{
    gboolean forward = TRUE;

    if (delta==0)
        return current_offset;

    if (delta<0)
    {
        delta = abs(delta);
        forward = FALSE;
    }
    while (delta--)
    {
        offset_type temp;

        if (forward)
            temp = wrap_get_eol(dp, current_offset);
        else
            temp = find_previous_wrapped_text_line(dp, current_offset);


        // Offset didn't changed ? we've reached eof
        if (temp==current_offset)
            break;

        current_offset = temp;
    }

    return current_offset;
}


static offset_type wrap_get_eol(GVDataPresentation *dp, offset_type start_of_line)
{
    offset_type offset;
    char_type value;

    /* A Single TAB character in the file,
       Translates to several displayable characters on the screen.
       We need to take that into account when calculating number of
       characters before wraping the line */
    guint char_count = 0;

    offset = start_of_line;

    while (TRUE)
    {
        value = gv_input_mode_get_utf8_char(dp->imd, offset);

        if (value==INVALID_CHAR)
            break;

        offset = gv_input_get_next_char_offset(dp->imd, offset);

        // break upon end of line
        if (value=='\n' || value=='\r')
            break;

        if (value=='\t')
            char_count += dp->tab_size;
        else
            char_count++;

        if (char_count >= dp->wrap_limit)
            break;
    }

    return offset;
}


static offset_type binfixed_align_offset(GVDataPresentation *dp, offset_type offset)
{
    offset_type o;

    g_return_val_if_fail (dp->fixed_count>0, offset);

    if (offset > dp->max_offset)
        offset = dp->max_offset;
    o = ((offset_type)(offset / dp->fixed_count) * dp->fixed_count);

    return o;
}


static offset_type binfixed_scroll_lines(GVDataPresentation *dp, offset_type current_offset, int delta)
{
    g_return_val_if_fail (dp->fixed_count>0, current_offset);

    if (delta > 0)
    {
        if (current_offset + delta * dp->fixed_count > dp->max_offset)
            delta = (dp->max_offset-current_offset) / dp->fixed_count;

        return delta * dp->fixed_count + current_offset;
    }
    else
    {
        delta = abs(delta);

        if (delta * dp->fixed_count > current_offset)
            return 0;

        return current_offset - delta * dp->fixed_count;
    }
}


static offset_type binfixed_get_eol(GVDataPresentation *dp, offset_type start_of_line)
{
    g_return_val_if_fail (dp->fixed_count>0, start_of_line);

    if (start_of_line + dp->fixed_count > dp->max_offset)
        return dp->max_offset;

    return start_of_line + dp->fixed_count;
}

File: src/intviewer/data_presentation.rs
// SPDX-FileCopyrightText: 2025 Andrey Kutejko <andy128k@gmail.com>
// SPDX-FileCopyrightText: 2026 Wladimir Palant https://palant.info/
//
// SPDX-License-Identifier: GPL-3.0-or-later

use crate::intviewer::input_modes::InputMode;

pub struct DataPresentation {
    mode: DataPresentationMode,
    wrap_limit: u32,
    tab_size: u32,
    fixed_count: u32,
}

impl DataPresentation {
    pub fn new() -> Self {
        Self {
            mode: DataPresentationMode::NoWrap,
            wrap_limit: 80,
            tab_size: 8,
            fixed_count: 16,
        }
    }

    #[inline]
    pub fn mode(&self) -> DataPresentationMode {
        self.mode
    }

    #[inline]
    pub fn set_mode(&mut self, mode: DataPresentationMode) {
        self.mode = mode;
    }

    pub fn align_offset_to_line_start(&self, input_mode: &InputMode, offset: u64) -> u64 {
        match self.mode() {
            DataPresentationMode::NoWrap => self.nowrap_align_offset(input_mode, offset),
            DataPresentationMode::Wrap => self.wrap_align_offset(input_mode, offset),
            DataPresentationMode::BinaryFixed => self.binfixed_align_offset(input_mode, offset),
        }
    }

    pub fn scroll_lines(&self, input_mode: &InputMode, current_offset: u64, delta: i32) -> u64 {
        match self.mode() {
            DataPresentationMode::NoWrap => {
                self.nowrap_scroll_lines(input_mode, current_offset, delta)
            }
            DataPresentationMode::Wrap => self.wrap_scroll_lines(input_mode, current_offset, delta),
            DataPresentationMode::BinaryFixed => {
                self.binfixed_scroll_lines(input_mode, current_offset, delta)
            }
        }
    }

    pub fn end_of_line_offset(&self, input_mode: &InputMode, start_of_line: u64) -> u64 {
        match self.mode() {
            DataPresentationMode::NoWrap => self.nowrap_get_eol(input_mode, start_of_line),
            DataPresentationMode::Wrap => self.wrap_get_eol(input_mode, start_of_line),
            DataPresentationMode::BinaryFixed => self.binfixed_get_eol(input_mode, start_of_line),
        }
    }

    pub fn set_tab_size(&mut self, tab_size: u32) {
        self.tab_size = tab_size;
    }

    pub fn set_fixed_count(&mut self, chars_per_line: u32) {
        self.fixed_count = chars_per_line;
    }

    pub fn set_wrap_limit(&mut self, chars_per_line: u32) {
        self.wrap_limit = chars_per_line;
    }

    /// scans the file from offset "start" backwards, until a CR/LF is found.
    /// returns the offset of the previous CR/LF, or 0 (if we've reached the start of the file)
    fn find_previous_crlf(&self, input_mode: &InputMode, start: u64) -> Option<u64> {
        let mut offset = start;
        while offset > 0 {
            offset = input_mode.previous_char_offset(offset);
            match input_mode.character(offset) {
                None => return None,
                Some('\r') => return Some(offset),
                Some('\n') => {
                    let previous_offset = input_mode.previous_char_offset(offset);
                    if input_mode.character(previous_offset) == Some('\r') {
                        // Treat CRLF combination as a single EOL
                        return Some(previous_offset);
                    } else {
                        return Some(offset);
                    }
                }
                _ => {}
            }
        }
        None
    }

    fn nowrap_align_offset(&self, input_mode: &InputMode, mut offset: u64) -> u64 {
        while offset > 0 {
            let prev_offset = input_mode.previous_char_offset(offset);
            match input_mode.character(prev_offset) {
                None => return 0,
                Some('\n') | Some('\r') => return offset,
                _ => offset = prev_offset,
            }
        }
        0
    }

    fn nowrap_scroll_lines(
        &self,
        input_mode: &InputMode,
        mut current_offset: u64,
        mut delta: i32,
    ) -> u64 {
        if delta == 0 {
            return current_offset;
        }

        let mut forward = true;
        if delta < 0 {
            delta = -delta;
            forward = false;
        }

        for _ in 0..delta {
            let temp = if forward {
                self.nowrap_get_eol(input_mode, current_offset)
            } else {
                let Some(temp) = self
                    .find_previous_crlf(input_mode, current_offset)
                    .map(|t| self.nowrap_align_offset(input_mode, t))
                else {
                    break;
                };
                temp
            };

            // Offset didn't changed ? we've reached eof
            if temp == current_offset {
                break;
            }

            current_offset = temp;
        }
        current_offset
    }

    fn nowrap_get_eol(&self, input_mode: &InputMode, start_of_line: u64) -> u64 {
        let mut offset = start_of_line;
        loop {
            let Some(value) = input_mode.character(offset) else {
                break;
            };

            offset = input_mode.next_char_offset(offset);

            // break upon end of line
            if value == '\n' {
                break;
            } else if value == '\r' {
                // Treat CRLF combination as a single EOL
                if let Some(value) = input_mode.character(offset)
                    && value == '\n'
                {
                    offset = input_mode.next_char_offset(offset);
                }
                break;
            }
        }
        offset
    }

    /// returns the start offset of the previous line,
    /// with special handling for wrap mode.
    fn find_previous_wrapped_text_line(&self, input_mode: &InputMode, start: u64) -> u64 {
        // step 1: find start offset of previous text line
        let Some(mut offset) = self
            .find_previous_crlf(input_mode, start)
            .map(|t| self.nowrap_align_offset(input_mode, t))
        else {
            return start;
        };

        // Step 2
        loop {
            let next_line_offset = self.wrap_get_eol(input_mode, offset);

            // this is the line we want: When the next line's offset is the current
            // offset ('start' parameter), 'offset' will point to the previous
            // displayable line

            if next_line_offset >= start {
                return offset;
            }

            offset = next_line_offset;
        }
    }

    pub fn wrap_align_offset(&self, input_mode: &InputMode, offset: u64) -> u64 {
        let mut line_start = self.nowrap_align_offset(input_mode, offset);
        let mut temp = line_start;
        while temp <= offset {
            line_start = temp;
            temp = self.wrap_scroll_lines(input_mode, temp, 1)
        }
        line_start
    }

    fn wrap_scroll_lines(
        &self,
        input_mode: &InputMode,
        mut current_offset: u64,
        mut delta: i32,
    ) -> u64 {
        if delta == 0 {
            return current_offset;
        }

        let mut forward = true;
        if delta < 0 {
            delta = -delta;
            forward = false;
        }

        for _ in 0..delta {
            let temp = if forward {
                self.wrap_get_eol(input_mode, current_offset)
            } else {
                self.find_previous_wrapped_text_line(input_mode, current_offset)
            };

            // Offset didn't change? we've reached eof
            if temp == current_offset {
                break;
            }

            current_offset = temp;
        }
        current_offset
    }

    fn wrap_get_eol(&self, input_mode: &InputMode, start_of_line: u64) -> u64 {
        /* A Single TAB character in the file,
        Translates to several displayable characters on the screen.
        We need to take that into account when calculating number of
        characters before wraping the line */
        let mut char_count = 0;

        let mut offset = start_of_line;

        loop {
            let Some(value) = input_mode.character(offset) else {
                break;
            };

            offset = input_mode.next_char_offset(offset);

            // break upon end of line
            if value == '\n' {
                break;
            } else if value == '\r' {
                // Treat \r\n combination as a single EOL
                if let Some(value) = input_mode.character(offset)
                    && value == '\n'
                {
                    offset = input_mode.next_char_offset(offset);
                }
                break;
            } else if value == '\t' {
                char_count = next_tab_position(char_count, self.tab_size);
            } else {
                char_count += 1;
            }

            if char_count >= self.wrap_limit {
                break;
            }
        }
        offset
    }

    fn binfixed_align_offset(&self, input_mode: &InputMode, offset: u64) -> u64 {
        let max_offset = input_mode.max_offset();
        let fixed_count = self.fixed_count.max(1) as u64;
        offset.clamp(0, max_offset) / fixed_count * fixed_count
    }

    fn binfixed_scroll_lines(
        &self,
        input_mode: &InputMode,
        current_offset: u64,
        delta: i32,
    ) -> u64 {
        let fixed_count = self.fixed_count.max(1);
        (self
            .binfixed_align_offset(input_mode, current_offset)
            .saturating_add_signed(fixed_count as i64 * delta as i64))
        .clamp(0, input_mode.max_offset())
    }

    fn binfixed_get_eol(&self, input_mode: &InputMode, start_of_line: u64) -> u64 {
        self.binfixed_scroll_lines(input_mode, start_of_line, 1)
    }
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, strum::FromRepr, Default)]
#[repr(C)]
pub enum DataPresentationMode {
    #[default]
    NoWrap = 0,
    Wrap,
    /// fixed number of binary characters per line.
    /// e.g. CHAR=BYTE, no UTF-8 or other translations
    BinaryFixed,
}

pub fn next_tab_position(column: u32, tab_size: u32) -> u32 {
    (column + tab_size) / tab_size * tab_size
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn set_data_presentation_mode() {
        let mut dp = DataPresentation::new();
        for mode in [
            DataPresentationMode::Wrap,
            DataPresentationMode::NoWrap,
            DataPresentationMode::BinaryFixed,
        ] {
            dp.set_mode(mode);
            assert_eq!(dp.mode(), mode);
        }
    }

    const ABRACADABRA: &[u8] = br#"
a
ab
abra
abrac
abraca
abracad
abracada
abracadab
abracadabr
abracadabra
"#;

    #[test]
    fn nowrap() {
        let imd = InputMode::new(ABRACADABRA);
        let mut dp = DataPresentation::new();
        dp.set_mode(DataPresentationMode::NoWrap);

        assert_eq!(dp.nowrap_align_offset(&imd, 0), 0);
        assert_eq!(dp.nowrap_align_offset(&imd, 1), 1);
        assert_eq!(dp.nowrap_align_offset(&imd, 2), 1);
        assert_eq!(dp.nowrap_align_offset(&imd, 3), 3);
        assert_eq!(dp.nowrap_align_offset(&imd, 4), 3);
        assert_eq!(dp.nowrap_align_offset(&imd, 5), 3);
        assert_eq!(dp.nowrap_align_offset(&imd, 6), 6);

        assert_eq!(dp.scroll_lines(&imd, 0, 1), 1);
        assert_eq!(dp.scroll_lines(&imd, 1, 1), 3);
        assert_eq!(dp.scroll_lines(&imd, 3, 1), 6);

        assert_eq!(dp.scroll_lines(&imd, 6, -1), 3);
        assert_eq!(dp.scroll_lines(&imd, 3, -1), 1);
        assert_eq!(dp.scroll_lines(&imd, 1, -1), 0);
    }

    #[test]
    fn test_next_tab_position() {
        assert_eq!(next_tab_position(0, 8), 8);
        assert_eq!(next_tab_position(1, 8), 8);
        assert_eq!(next_tab_position(7, 8), 8);
        assert_eq!(next_tab_position(8, 8), 16);
        assert_eq!(next_tab_position(9, 8), 16);
    }
}
